src/main/java/org/jenkinsci/plugins/bitbucket_approve/BitbucketApprover.java

```diff
@@ -81,32 +85,107 @@ public boolean perform(AbstractBuild build, Launcher launcher, BuildListener lis
             return false;
         }
 
-        String eOwner = build.getEnvironment(listener).expand(mOwner);
-        String eSlug  = build.getEnvironment(listener).expand(mSlug);
-        String url = String.format("https://api.bitbucket.org/2.0/repositories/%s/%s/commit/%s/approve", eOwner, eSlug, commitHash);
-        logger.println("Bitbucket Approve: " + url);
-
         OkHttpClient client = new OkHttpClient();
         client.setConnectTimeout(30, TimeUnit.SECONDS);
         client.setReadTimeout(60, TimeUnit.SECONDS);
 
-        Request.Builder builder = new Request.Builder();
-        Request request = builder.header("Authorization", getDescriptor().getBasicAuth())
-                .url(url)
-                .method("POST", null).build();
+        String buildState = (build.getResult().ordinal == Result.SUCCESS.ordinal) ? "SUCCESSFUL" : "FAILED";
+
+        if ((build.getResult().ordinal == Result.UNSTABLE.ordinal && !mApproveUnstable)
+             || (build.getResult().isWorseThan(Result.UNSTABLE))) {
+
+            if (!mApprovalMethod.equals("statusOnly")) {
+                logger.println("Bitbucket Approve: Skipping approval because build is " + build.getResult().toString());
+            }
+
+            return postBuildStatus(build, logger, client, commitHash, buildState);
+        }
+
+        boolean result = true;
+
+        if (!mApprovalMethod.equals("statusOnly")) {
+            String eOwner = build.getEnvironment(listener).expand(mOwner);
+            String eSlug  = build.getEnvironment(listener).expand(mSlug);
+            String url = String.format("https://api.bitbucket.org/2.0/repositories/%s/%s/commit/%s/approve", eOwner, eSlug, commitHash);
+            logger.println("Bitbucket Approve: " + url);
+
+            Request.Builder builder = new Request.Builder();
+            Request request = builder.header("Authorization", getDescriptor().getBasicAuth())
+                    .url(url)
+                    .method("POST", null).build();
+
+            try {
+                Response response = client.newCall(request).execute();
+
+                if (!isSuccessful(response)) {
+                    result = false;
+                    logger.println("Bitbucket Approve: " + response.code() + " - " + response.message());
+                    logger.println("Bitbucket Approve: " + response.body().string());
+                }
+
+
+            } catch (IOException e) {
+                e.printStackTrace(listener.getLogger());
+            }
+        }
+
+        if (!postBuildStatus(build, logger, client, commitHash, buildState)) {
+            result = false;
+        }
+
+        return result;
+    }
+
+    private boolean postBuildStatus(AbstractBuild build, PrintStream logger, OkHttpClient client, String commitHash, String state) {
+        if (mApprovalMethod.equals("approveOnly")) {
```

Comment:
Condition checked on a different level of abstraction than the corresponding `if (!mApprovalMethod.equals("statusOnly"))`. 

I think it would be more clear if we had two methods, one for updating the status and one for approving. Within the `perform()` method we should take the decision whether or not to call them.